### ISSUE TITLE
Sandbox should manipulate response headers in the same way API Gateway does

### ISSUE THREAD

**Issue by andybee** [architect/architect#879]
Recently put some code in to an Architect project that relied on sending certain HTTP headers. Worked fine in Sandbox, but once deployed to AWS realised the header I was relying on is manipulated by API Gateway. This behaviour is [documented by AWS](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-known-issues.html).

Whilst this falls firmly in the "nice to have", ideally Sandbox would closely mimic the behaviour of API Gateway so these things are caught earlier on in the development cycle.

I note that a [commit in March](https://github.com/architect/sandbox/commit/818c4bb2446ce2a03a7e1795a1a170f3078a7b0b) implemented this behaviour for incoming request headers, but not response headers. Not sure if that same function should be used for both, or separate strategies should be used for request and response. Because the behaviour does differ I feel it should be more explicit on request vs response.

I've produced a demo repo that has a single endpoint which sets all of the "dropped" or "remapped" headers.

https://github.com/andybee/arc-apigateway-header-manipulation-demo

**Comment by ryanblock** [comment]
I find it genuinely pretty neat that after years of building Architect we're still finding new surfaces in AWS that we haven't covered. This is a big one and I definitely do not consider it nice to have – this is absolutely core requirement for Sandbox. Please stand by!

**Comment by ryanblock** [comment]
Ok, fix for this should be live now in `v1.12.3-RC.0` (`npm i @architect/sandbox@RC`). Worth noting: I believe the comments in [your example](https://github.com/andybee/arc-apigateway-header-manipulation-demo/blob/main/src/http/get-index/index.js) follow the API Gateway issue above, however the behavior I actually observed in what gets dropped and remapped differs from their docs.

I only tested those headers against a live REST API (for now), not HTTP, but here's what I saw remapped and dropped: https://github.com/architect/sandbox/blob/header-formatting-round-2/src/http/invoke-http/_res-header-fmt.js#L9-L27

Please to check that, though, and let me know if I've somehow got this wrong!

Full PR here for anyone interested: https://github.com/architect/sandbox/pull/393

**Comment by andybee** [comment]
Looks great. I can't vouch for all the header behaviours, but the specific issue I hit is handled by this update. Presumably you deploys my demo repo and mocked the behaviour of that within AWS in to Sandbox?

Thanks for picking this up, the reason I originally pinned this as "nice to have" is that I could understand an argument that it's hard to endlessly monitor what the current state of AWS deployment is likely to be and continue to keep Sandbox up-to-date, but if you're willing to try that it's great because catching these issues early on during local development helps a lot of people fall down a blackhole of "this is great, I'll just deploy..." and then hitting a brick wall.

**Comment by ryanblock** [comment]
It is hard to monitor the current state of AWS, but I'd also argue that's a core goal of the project – Architect exists to make this stuff easier, and we can't do that unless we do the hard work of making the local dev experience as close to perfect as possible. The scenario you described is exactly the scenario we strive to avoid!

So on that note, what I wound up doing was working through every header mentioned in the [headers table of that API Gateway doc](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-known-issues.html). I called the function with every header (and logged the results), responded with every header, and cataloged the results in a spreadsheet.

Here's the function I used:
```
exports.handler = async function http (req) {
  console.log(`req:`, req)
  return {
    body: 'here are all the headers',
    headers: {
      'Accept-Charset': 'Accept-Charset',
      'Accept-Encoding': 'Accept-Encoding',
      'Accept': 'Accept',
      'Age': 'Age',
      'Authorization': 'Authorization',
      'Connection': 'Connection',
      'Content-Encoding': 'identity',
      'Content-Length': '0',
      'Content-MD5': 'Content-MD5',
      'Content-Type': 'Content-Type',
      'Date': 'Date',
      'Expect': 'Expect',
      'Host': '96wrgyfsba.execute-api.us-east-1.amazonaws.com',
      'Max-Forwards': 'Max-Forwards',
      'Pragma': 'Pragma',
      'Proxy-Authenticate': 'Proxy-Authenticate',
      'Range': 'Range',
      'Referer': 'Referer',
      'Server': 'Server',
      'TE': 'trailers',
      'Trailer': 'Trailer',
      'Transfer-Encoding': 'Transfer-Encoding',
      'Upgrade': 'Upgrade',
      'User-Agent': 'User-Agent',
      'Via': 'Via',
      'Warn': 'Warn',
      'WWW-Authenticate': 'WWW-Authenticate',
      'Set-Cookie': 'foo',
      'control': 'control',
    }
  }
}
```
Here are the results (tested on an Arc 6 REST API). Note on header case changes: bold indicates forced upper Pascal-kebab case, otherwise everything gets normalized to lowcase.
|Header|Request|Response|Remapped responses|
|-|-|-|-|
|Accept-Charset|accept-charset|accept-charset||
|Accept-Encoding|accept-encoding|accept-encoding||
|Accept|accept|accept||
|Age|age|age||
|Authorization|**Authorization**|-REMAPPED-|x-amzn-remapped-authorization|
|Connection|-DROPPED-|-REMAPPED-|x-amzn-remapped-connection|
|Content-Encoding|Content-Encoding|content-encoding||
|Content-Length|-DROPPED-|content-length|x-amzn-remapped-content-length|
|Content-MD5|-DROPPED-|-REMAPPED-|x-amzn-remapped-content-md5|
|Content-Type|content-type|content-type||
|Date|**Date**|date|x-amzn-remapped-date|
|Expect|-DROPPED-|-REMAPPED-|x-amzn-remapped-expect|
|Host|**Host**|-REMAPPED-|x-amzn-remapped-host|
|Max-Forwards|-DROPPED-|-REMAPPED-|x-amzn-remapped-max-forwards|
|Pragma|pragma|pragma||
|Proxy-Authenticate|-DROPPED-|-REMAPPED-|x-amzn-remapped-proxy-authenticate|
|Range|range|range||
|Referer|referer|referer||
|Server|-DROPPED-|-REMAPPED-|x-amzn-remapped-server|
|TE|-DROPPED-|te||
|Trailer|-DROPPED-|-REMAPPED-|x-amzn-remapped-trailer|
|Transfer-Encoding|-DROPPED-|-DROPPED-||
|Upgrade|-DROPPED-|-REMAPPED-|x-amzn-remapped-upgrade|
|User-Agent|**User-Agent**|-REMAPPED-|x-amzn-remapped-user-agent|
|Via|via|via||
|Warn|warn|warn||
|WWW-Authenticate|-DROPPED-|-REMAPPED-|x-amzn-remapped-www-authenticate|

**Comment by ryanblock** [comment]
Just for reference, here's the actual headers logged (copied out of CloudWatch):

**Comment by andybee** [comment]
Trying to work out if there’s value in trying to automate this in some way? Like an automated test that could run periodically and alert us to a change?

Big issue I can see with that is knowing which headers to observe and the fact that regardless of which standards we base it on, the world keeps turning and new headers will come along that aren’t covered - thus meaning the test gives a false sense of security (“Everything’s fine! Until you use that new header we’ve not covered and then it’s broken again!”)

**Comment by ryanblock** [comment]
I think there's definitely value in having that live infra integration test, and I'd welcome any PRs for that! Ideally it would cover every header in HTTP specs, which I imagine would minimize the scenario you mentioned. That said, I don't think it's super urgent, so we should be good for a minute.

Ok, so I'm going to merge this!

**Comment by andybee** [comment]
Excellent, thanks!

I've been looking for an excuse/reason to have a look at GitHub Actions, so this might be a good project to dip my toes in the water. I'll give it a think.

**Comment by ryanblock** [comment]
Shipped in Sandbox `1.12.3`, and should be released in Architect `6.5` this week!
